File: cpp/odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (!head) return NULL;
        
        ListNode *odd = head, *even = head->next, *l = odd, *r = even, *pl;
        
        bool isOdd = true;
        while (r) {
            l->next = r->next;
            pl = l;
            l = r;
            r = r->next;
            isOdd = !isOdd;
        }
        
        if (isOdd) l->next = even;
        else pl->next = even;
        
        return head;
    }
};

File: go/odd-even-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func oddEvenList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	odd := head
	even := head.Next

	var pl *ListNode
	l := odd
	r := even
	isOdd := true
	for r != nil {
		l.Next = r.Next
		pl = l
		l = r
		r = r.Next
		isOdd = !isOdd
	}

	if isOdd {
		l.Next = even
	} else {
		pl.Next = even
	}

	return head
}